### PR TITLE
fix(replay): verify mask alignment instead of discarding frames on any redraw

### LINKED ISSUE #596: Some screens appear to be dropped in session replay recordings
## Description

Some screens appear to be dropped in session replay recordings on Android. Certain screens that the user navigates to during a session do not appear in the resulting replay, leaving gaps in the recorded flow.

Ticket: https://posthoghelp.zendesk.com/agent/tickets/60852 (moved to PostHog: https://us.posthog.com/project/2/support/tickets/63955)

### PR DESCRIPTION
**Pull request by arnohillen**
## :bulb: Motivation and Context

Closes #596.

> [!NOTE]
> The new mask-alignment verification is opt-in and defaults off. Enable it with `config.sessionReplayConfig.verifyScreenshotMaskAlignment = true`; otherwise the existing screenshot redraw guard remains in use.

In screenshot mode, a frame was discarded whenever its window redrew during the PixelCopy capture, unless an animation-type heuristic matched (`hasTransientState` from #529, surface/texture views from #649). Most animated content matches neither signal: indeterminate spinners (`ProgressDialog`), animated GIFs (Glide/Coil), Lottie (which never sets transient state, so the #529 exemption structurally could not fire for it), Material progress indicators, and Compose infinite animations all redraw per frame on the UI thread. On screens showing any of them, essentially every capture logged "Session Replay screenshot discarded due to screen changes" and the replay showed nothing. On top of that, the draw-dirty flags were single fields shared across all tracked windows, so an animating loader dialog also blanked captures of the static activity behind it.

The guard exists for a real reason (#254 / #234): mask rects are computed from live views after the pixels are frozen, so a structural change mid-capture can drift masks off sensitive content. This PR keeps that protection but stops proxying it with "did anything redraw":

- **Per-window draw state**: `isOnDrawnCalled`/`didLayoutSinceReset` move into a `WindowDrawState` on `ViewTreeSnapshotStatus`. PixelCopy copies a single window's surface and masks come from that window's own tree, so one window's draws say nothing about another window's mask alignment.
- **Verified masks instead of heuristics**: mask rects are sampled before `PixelCopy.request` and again in the callback. A dirty frame is kept only when both walks agree, no layout pass ran, and neither walk was poisoned. Pixel-only animation redraws pass this check no matter which library drives them; structural changes still discard. The `hasTransientState`/surface-view exemptions are deleted: frames they legitimately kept have stable geometry and pass rect equality anyway, and frames they kept with moving masked geometry were unsafe to keep at all.
- **Fail closed on unknowable geometry**: a walk is poisoned (frame discarded) when it meets a rendered view it cannot place (legacy `view.animation`, transient state mid-animation), and when the Compose semantics pass times out. Previously such views were silently pruned from the walk, which would have shipped them unmasked. A timed-out PixelCopy latch also no longer ships the bitmap before masks are painted.

Behavior is monotone for safety: no frame that was previously discarded for a genuine structural change is now kept, and the discard log line is unchanged for support diagnostics. The default wireframe mode is untouched, and the Flutter/RN forced-screenshot bridge inherits the fix through the same path.

## :green_heart: How did you test it?

- Unit tests for the keep/discard predicate: clean frame, stable rects, moved rects, appeared/disappeared rects, layout pass, poisoned walks.
- Deterministic end-to-end Robolectric tests through `generateSnapshot` + `ShadowPixelCopy`, using a hook view that injects state changes exactly between the pre- and post-copy walks: dirty-but-stable frame kept (the fix path), masked widget moved mid-capture discarded, layout mid-capture discarded, masked view mid legacy animation fails closed (regression test for the walk-pruning hole), and a redraw+layout in another window no longer discards this window's capture (fails under the old shared-flag code).
- Full `posthog-android` unit test suite, `apiCheck`, and `spotlessCheck` pass locally.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## :zap: Performance

The verification machinery sits on two hot paths: the endpoint walks (twice per ~1s capture) and the draw-time walk, which runs on the main thread on **every frame** while a capture is in flight, on exactly the animated screens this PR unblocks. Follow-up commits make those paths allocation-free and add fail-fast exits, with discard semantics unchanged:

- The draw-time walk streams each mask rect against the capture baseline and stops on the first mismatch, instead of building a `List<Rect>` per frame and comparing it under the lock.
- Walk-confined scratch `Rect`/`Point` objects replace the per-view `Rect()`/`Point()` allocations, and a primitive int set replaces the boxing `MutableSet<Int>` for cycle detection.
- Doomed captures fail fast: a draw or layout during the pre-walk skips the ~10MB ARGB_8888 bitmap and PixelCopy entirely (with a bounded re-arm so a busy screen still captures); once a capture's verdict is sealed as discard, per-frame sampling and the post-copy walk stop.
- The per-frame draw callback is lock-free (a monotone draw counter, main thread is the single writer) and allocation-free: ~7-10 ns when no capture is in flight.
- The redundant `rootView` attach check (verified against AOSP) and the per-walk `toString()` copies of TextView text/hint are gone.

**What the numbers show.** Two committed benchmarks, both runnable with `POSTHOG_BENCHMARK=1 ./gradlew :posthog-android:testReleaseUnitTest --tests '<name>'` (they never run in CI):

`PostHogReplayMaskWalkMicroBenchmark` isolates the changed data structures (median of 5 batches x 500k ops, 3 runs; bytes/op are exact per-thread allocation counts):

| Primitive | Before | After |
|---|---|---|
| Visited-set, 60 views | 568 ns, 3,936 B | 233 ns, 0 B (2.4x faster) |
| Visited-set, 550 views | 5.8 µs, 34,688 B | 2.0 µs, 16 B (2.9x faster) |
| Rect compare, matching frame (40 rects) | 310 ns, 1,456 B | 326 ns, 0 B (same speed, zero alloc) |
| Rect compare, mismatching frame (40 rects) | 306 ns, 1,456 B | 21 ns, 0 B (15x faster, early exit) |

`PostHogReplayMaskWalkBenchmark` runs the full walk on a real 60/550-view tree. An A/B comparison (4 alternating runs against the pre-optimization commit) shows **wall-clock parity within noise there**: on Robolectric the walk is dominated by shadow-framework calls both versions make identically. That benchmark's value is the allocation counter and as a regression harness, not as proof of speedup.

So the honest claim is: on device, the SDK's own per-frame cost during capture windows (one `Rect`+`Point` per visited view, one boxed `Integer`+map entry per view, one stored `Rect` per masked view, a full-list comparison under a lock, ~KB-scale garbage per frame at 60-120 fps) drops to zero allocation and less work per view, mismatching frames stop at the first divergent rect, and unkeepable captures skip the bitmap+PixelCopy+encode entirely. Framework-call count per walk is unchanged.

## :white_check_mark: Correctness evidence

All of this runs in CI on every push:

- **Exhaustive protocol check** (`WindowDrawStateProtocolTest`): every placement of up to three mutate/draw/layout events across the capture pipeline's five gaps (1,096 schedules, enumerated, count asserted) runs against the real `WindowDrawState`. Kept frames provably never coexist with a layout, a draw overlapping the pre-walk, or a draw whose geometry differs from the baseline; stable-geometry schedules (pixel-only animations, the bug this PR fixes) are provably always kept.
- **Property tests** (`PostHogReplayMaskWalkPropertyTest`): the streaming comparison returns exactly the same verdict as the old build-a-list-and-compare across 10,000 randomized baseline mutations (element changed / inserted / removed / swapped / truncated); the primitive int set matches a `HashSet` oracle across 55,000 operations including the zero-sentinel edge case and growth.
- **Behavioral tests**: 280 unit tests total, including end-to-end Robolectric captures pinning every keep/discard scenario (stable redraw kept; moved/appeared/disappeared masks discarded; layout discarded; poisoned walks fail closed; pre-walk draws re-arm then capture; in-flight draw samples fail closed; per-window isolation).
- **Adversarial review**: the capture protocol went through three rounds of independent adversarial verification (multiple reviewers per round, prompted to construct PII-leak interleavings against the real code). The one design they refuted (keeping pre-walk-overlapping frames when rect geometry agrees) was discarded; the shipped protocol is the one that survived.

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

- Authored with PostHog Code (Claude). Root cause was established with a multi-agent investigation over the SDK source, AOSP internals (`ProgressBar`/`AnimatedVectorDrawable`/`ViewRootImpl`), and the fix history of #254/#529/#649, then adversarially verified claim by claim.
- Notable findings that shaped the design: `setHasTransientState` is only called by `ViewPropertyAnimator`, `Editor`, `Transition`, and view-translation in the framework, so the #529 exemption never fired for Lottie or any drawable-level animation; sibling SDKs (posthog-ios, Sentry Android) avoid this bug class by sampling masks synchronously with the frame, which is what the rect pre/post verification approximates while keeping PixelCopy off the main thread.
- An enumerate-more-animation-types fix (like closed PR #606) was considered and rejected: it cannot cover Compose or future libraries and keeps unverified keep paths. Rect verification subsumes the heuristics.
- The diff itself was adversarially reviewed by independent agents (PII, concurrency, regression, efficacy lenses); the confirmed findings (legacy-animation walk pruning, Compose walk timeout, stale exemption flag) are fixed and regression-tested in this PR.

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

**Comment by marandaneto** [comment]
@arnohillen, this requires manual testing; otherwise, we risk leaking PII. Have you tested this, or are you purely relying on the unit tests?

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 1]
```diff
@@ -1202,12 +1146,42 @@ public class PostHogReplayIntegration(
...
+        preWalk: MaskWalk,
+        postWalk: MaskWalk,
+    ): Boolean {
+        if (preWalk.poisoned || postWalk.poisoned) {
+            return false
+        }
+        if (!drawState.isOnDrawnCalled) {
+            return true
+        }
+        return !drawState.didLayoutSinceReset &&
```
Comment (marandaneto) [inline posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: **[P1] Equal endpoint masks do not prove copy-time alignment.** A masked view can move from rect A to B, be captured by PixelCopy at B, then return to A before this comparison. Translation and Compose motion need not trigger global layout, so the equal A endpoints accept the frame while sensitive pixels remain unmasked at B. Any redraw involving movable masks needs stronger synchronization than endpoint equality.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 2]
```diff
@@ -1202,12 +1146,42 @@ public class PostHogReplayIntegration(
...
+    // Lottie, Compose) be captured at all instead of being discarded on every attempt.
+    internal fun shouldKeepFrame(
+        drawState: WindowDrawState,
+        preWalk: MaskWalk,
+        postWalk: MaskWalk,
+    ): Boolean {
+        if (preWalk.poisoned || postWalk.poisoned) {
+            return false
+        }
+        if (!drawState.isOnDrawnCalled) {
```
Comment (marandaneto) [inline posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: **[P1] Do not bypass alignment checks before `onDraw` fires.** Layout or mask geometry can change after pixels are frozen but before the next draw callback; in that interval this branch accepts the frame even when `didLayoutSinceReset` is true or the walks differ, then paints post-change masks over pre-change pixels. The clean path should still reject layouts and mismatched walks.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 3]
```diff
@@ -1160,11 +1096,19 @@ public class PostHogReplayIntegration(
...
-            config.logger.log("Session Replay findMaskableComposeWidgets failed: $e")
+        val completed =
+            try {
+                // await for 1s max
+                latch.await(1000, TimeUnit.MILLISECONDS)
+            } catch (e: Throwable) {
+                config.logger.log("Session Replay findMaskableComposeWidgets failed: $e")
+                false
+            }
+        if (completed) {
```
Comment (marandaneto) [inline posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: **[P1] Poison Compose walks when semantics traversal fails.** `completed` only means the posted runnable ended: the `RootForTest` early return and caught semantics exceptions also count down the latch. Both walks can therefore become trustworthy empty lists and ship a redrawing Compose screen without masks. Mark the walk poisoned for every traversal failure, not only timeout or interruption.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 4]
```diff
@@ -496,6 +417,7 @@ public class PostHogReplayIntegration(
         status.sentMetaEvent = false
         status.keyboardVisible = false
         status.lastSnapshot = null
+        status.drawState.reset()
```
Comment (marandaneto) [inline posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: **[P1] Do not reset capture tracking with snapshot metadata.** `clearSnapshotStates()` can run while PixelCopy is pending, and this reset erases draw/layout evidence recorded since the pre-walk. The callback can then take the clean-frame path and paint current mask coordinates over older pixels. Either invalidate the pending capture or leave its draw state intact.

**Comment by dustinbyrne** [comment]
Verified: I ran the Android sample app in screenshot-based session replay mode with a continuously animated loading spinner visible. Before the patch, every capture attempt during the animation was discarded and no replay frames were sent; after the patch, frames were captured continuously with no discard errors, and the animation was represented by changing screenshot payloads.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 5]
```diff
@@ -1202,12 +1156,48 @@ public class PostHogReplayIntegration(
...
+    // kept have stable geometry and pass rect equality anyway, and frames they kept with moving
+    // masked geometry were unsafe to keep at all.
+    internal fun shouldKeepFrame(
+        drawState: WindowDrawState,
+        preWalk: MaskWalk,
+        postWalk: MaskWalk,
+    ): Boolean {
+        if (!drawState.isOnDrawnCalled) {
+            return true
+        }
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Couldn't validate this via test: the current PR head already rejects poisoned pre- and post-walks on the clean-frame path, and the targeted regression test passes.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 6]
```diff
@@ -1292,7 +1284,7 @@ public class PostHogReplayIntegration(
                     success = false
                 } finally {
                     // reset the draw-dirty flags since we've taken the screenshot
-                    resetDrawStateTracking()
+                    drawState.reset()
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Couldn't validate this via test: the current PR head protects newer capture state from stale callbacks, and the targeted regression test passes.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 1 above)
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 2 above)
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 3 above)
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 4 above)
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 7]
```diff
@@ -581,10 +503,10 @@ public class PostHogReplayIntegration(
             val decorViewsSnapshot = synchronized(decorViews) { decorViews.entries.map { it.toPair() } }
             decorViewsSnapshot.forEach { (view, status) ->
                 clearViewListeners(view, status)
+                status.drawState.reset()
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Couldn't validate this via test: stop/uninstall do reset the local draw flags, but their lifecycle and queue gates prevented the test from demonstrating that an unsafe frame can be emitted.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 8]
```diff
@@ -1078,33 +1009,38 @@ public class PostHogReplayIntegration(
...
                 }
 
-                if (!findMaskableWidgets(viewChild, maskableWidgets, visitedViews)) {
-                    // do not continue if the screen has changed
-                    return false
-                }
+                findMaskableWidgets(viewChild, walk, visitedViews)
             }
         }
+    }
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt:1123]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 9]
```diff
@@ -1237,62 +1211,60 @@ public class PostHogReplayIntegration(
...
-            // reset the draw-dirty flags since we are about to take a screenshot
-            resetDrawStateTracking()
+            drawState.reset()
+
+            // Sampled before the pixels freeze, so the post-copy walk can prove the mask
+            // geometry didn't move mid-capture.
+            val preWalk = MaskWalk()
+            synchronized(maskWalkLock) {
+                findMaskableWidgets(view, preWalk)
+            }
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 10]
```diff
@@ -172,8 +171,7 @@ public class PostHogReplayIntegration(
             color = Color.BLACK
         }
 
-    // Reusable objects to avoid per-view allocations during screenshot masking.
-    // Only accessed from the PixelCopy callback thread (or executor), so no synchronization needed.
+    // Shared scratch objects; every user must hold maskWalkLock.
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 11]
```diff
@@ -1078,33 +1009,38 @@ public class PostHogReplayIntegration(
...
         view: View,
-        maskableWidgets: MutableList<Rect>,
+        walk: MaskWalk,
     ) {
         val latch = CountDownLatch(1)
 
+        // Local list so a timed-out runnable that fires late cannot mutate the walk.
+        val maskableWidgets = mutableListOf<Rect>()
+
         // compose requires the handler to be on the main thread
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: Validated via test, will fix.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 12]
```diff
@@ -1078,33 +1009,38 @@ public class PostHogReplayIntegration(
...
-                    return false
-                }
-
                 val viewChild = view.getChildAt(i) ?: continue
 
                 if (!viewChild.isVisible()) {
+                    // A skipped-but-rendered view could be a masked widget we cannot place.
+                    if (viewChild.isRenderedButUnplaceable()) {
+                        walk.poisoned = true
+                    }
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt:1083]: Couldn't validate this via test: propagated transient state is rejected by the existing root visibility gate on both the pre-PR base and current head, so the claimed new regression did not reproduce.

**Comment by marandaneto** [comment]
> Verified: I ran the Android sample app in screenshot-based session replay mode with a continuously animated loading spinner visible. Before the patch, every capture attempt during the animation was discarded and no replay frames were sent; after the patch, frames were captured continuously with no discard errors, and the animation was represented by changing screenshot payloads.

awesome, @dustinbyrne @arnohillen worth testing this on react native as well since react native relies on that and sometimes there are some incompatibilities

**Comment by dustinbyrne** [comment]
working as expected on RN

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 13]
```diff
@@ -228,85 +223,25 @@ public class PostHogReplayIntegration(
...
+        val walk = MaskWalk()
+        try {
+            findMaskableWidgets(view, walk)
         } catch (e: Throwable) {
-            config.logger.log("Session Replay surface rendering check failed: $e.")
-            false
+            config.logger.log("Session Replay draw-time mask walk failed: $e.")
+            walk.poisoned = true
         }
+        drawState.recordMaskWalk(captureToken, walk.rects, walk.poisoned)
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: this is somewhat existential in that _this is the entire point of this change_

[the remark this answers is not in the document]
Comment (arnohillen) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: that's not good, I'm letting Claude do a performance review.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt [hunk 14]
```diff
@@ -1226,73 +1222,97 @@ public class PostHogReplayIntegration(
...
+        if (preWalk.poisoned) {
+            config.logger.log("Session Replay screenshot discarded due to screen changes.")
+            return null
+        }
+
+        val captureToken =
+            drawState.beginMaskCapture(
+                preWalk.rects,
+                drawGenerationBeforePreWalk,
+            )
```
[the remark this answers is not in the document]
Comment (arnohillen) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: [arno's agent] Partly disagree with the severity: the generation check only ever covered the pre-walk (~0.5-4ms of each ~1s attempt, benchmarked in PostHogReplayMaskWalkBenchmark), and draws during the rest of the capture are rect-verified, so animated screens capture on most attempts rather than never. Keeping pre-walk-overlapping frames on rect agreement was implemented and then rejected: three independent adversarial reviews each constructed the same interleaving where the pre-walk already reflects that draw's tree while PixelCopy freezes the frame before it, so agreement proves nothing about the shipped pixels. 26a5abc keeps the deterministic discard but arms detection before the pre-walk via a lock-free monotone draw counter (no per-frame lock, no sample/setBaseline race), and unkeepable captures now skip the bitmap and PixelCopy entirely.

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (arnohillen) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: [arno's agent] Addressed in 2ed86f8: the draw-time walk is allocation-free with a first-mismatch early exit and is skipped once a capture is doomed; per-frame verification must stay because PixelCopy freezes pixels at an unknown instant (measured cost is in the PR description).

File: posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (arnohillen) [reply posthog-android/src/main/java/com/posthog/android/replay/PostHogReplayIntegration.kt]: [arno's agent] Fixed in 26a5abc + 9eb46ac5: detection arms before the pre-walk via a lock-free draw counter, and a draw overlapping the pre-walk discards only that walk with a bounded re-arm (3 attempts), so pixel-only animation frames still arm while layout and poison discard immediately.

**Comment by dustinbyrne** [comment]
@arnohillen what do you think about putting this behind a configuration variable? my concern is that this could end up having a performance impact, especially on lower-end hardware or cases with an unusual amount of layout rects (just a guess, happy to hear push back on this).

if so, i'm happy to do the work!

**Comment by arnohillen** [comment]
> @arnohillen what do you think about putting this behind a configuration variable? my concern is that this could end up having a performance impact, especially on lower-end hardware or cases with an unusual amount of layout rects (just a guess, happy to hear push back on this).
> 
> if so, i'm happy to do the work!

Yes, I think that's a great idea! I'm not 100% certain of this PR so it would be good to be able to revert quickly in case of any issues.

**Comment by dustinbyrne** [comment]
> Yes, I think that's a great idea! I'm not 100% certain of this PR so it would be good to be able to revert quickly in case of any issues.

sounds good, i'll add that in. i've done a good amount of testing with this branch and the overall strategy seems sound. though, i'm not 100% on what we'd consider to promote this to be default on at the moment. we don't have any measurements we can rely on to definitively state that this is performant in all cases.

regarding rollback, yes, i agree - that is the scary part. there's really no way of rolling back mobile SDKs considering updating the SDK would require a somewhat lengthy resubmission process and the actual roll out to users can be slow / has a long tail.

**Comment by dustinbyrne** [comment]
Added this in eb03cd49. The new mask-alignment path is now opt-in via:

```kotlin
config.sessionReplayConfig.verifyScreenshotMaskAlignment = true
```

It defaults to `false`; while disabled, screenshot capture keeps the existing redraw guard and does not run the new draw-time mask walks. The capture mode is pinned for each in-flight screenshot so changing the mutable config mid-capture cannot mix the two protocols. I also added enabled/disabled and runtime-toggle coverage, updated the API dump and changeset, and ran the full `posthog-android` unit tests, `apiCheck`, and `spotlessCheck`.

**Review by dustinbyrne — APPROVED** [review]
worth taking another look through, but i believe this is good to go
